File: src/structured-light/GrayCode.cpp
#include "GrayCode.hpp"
#include <iostream>
#include <cmath>
#include <set>

GrayCode::GrayCode(cv::Size resolution, int axis, int px_f)
    : width(resolution.width), height(resolution.height), px_f(px_f){
    
    // Aloca as imagens: 2 canais extras baseados no original em Python
    n_bits = static_cast<int>(std::ceil(std::log2(static_cast<double>(width) / px_f) + 1.0));
    gc_images.resize(n_bits+2);

    for (int i = 0; i < n_bits+2; ++i) {
        gc_images[i] = cv::Mat::zeros(height, width, CV_8UC1);
    }
    // create_graycode_images();
}

void GrayCode::show_gc_image() const {
    for (size_t i = 0; i < gc_images.size(); ++i) {
        std::cout << i << std::endl;
        cv::imshow("Image", gc_images[i]);
        cv::waitKey(0);
    }
    cv::destroyAllWindows();
}

std::vector<std::string> GrayCode::list_to_graycode_binary(const std::vector<int>& int_list, int bit_length) const {
    std::vector<std::string> graycode_list;
    for (int n : int_list) {
        int graycode = n ^ (n >> 1);
        std::string binary_str = "";
        for (int k = bit_length - 1; k >= 0; --k) {
            binary_str += ((graycode >> k) & 1) ? "1" : "0";
        }
        graycode_list.push_back(binary_str);
    }
    return graycode_list;
}

void GrayCode::create_graycode_image() {
    gc_images[0].setTo(255); 
    gc_images[1].setTo(0); 

    std::vector<int> width_list;
    int max_val = 1 << n_bits; 
    int repeat_count = px_f / 2;
    
    for (int element = 0; element < max_val; ++element) {
        for (int r = 0; r < repeat_count; ++r) {
            if (width_list.size() < static_cast<size_t>(width)) {
                width_list.push_back(element);
            }
        }
    }
    
    while(width_list.size() < static_cast<size_t>(width)) {
        width_list.push_back(0);
    }

    std::vector<std::string> width_b = list_to_graycode_binary(width_list, n_bits);

    for (int j = 0; j < width; ++j) {
        for (int i = 0; i < n_bits; ++i) {
            uchar pixel_val = (width_b[j][i] == '1') ? 255 : 0;
            for (int h = 0; h < height; ++h) {
                gc_images[i + 2].at<uchar>(h, j) = pixel_val;
            }
        }
    }
}

std::vector<int> GrayCode::get_gc_order_v() const {
    std::vector<int> int_values(width, 0);
    
    for (int j = 0; j < width; ++j) {
        int current_val = 0;
        for (int i = 0; i < n_bits; ++i) {
            int bit = (gc_images[i + 2].at<uchar>(0, j) > 0) ? 1 : 0;
            current_val += bit * (1 << (n_bits - 1 - i));
        }
        int_values[j] = current_val;
    }

    std::vector<int> unique_vals;
    std::set<int> seen;
    for(int val : int_values) {
        if(seen.find(val) == seen.end()) {
            seen.insert(val);
            unique_vals.push_back(val);
        }
    }

    return unique_vals;
}

std::vector<cv::Mat> GrayCode::get_gc_images(const std::string& color) const {
    std::vector<cv::Mat> colored_images(gc_images.size());
    cv::Mat zero_channel = cv::Mat::zeros(height, width, CV_8UC1);

    for (size_t i = 0; i < gc_images.size(); ++i) {
        std::vector<cv::Mat> canais(3, zero_channel);
        
        if (color == "blue") {
            canais[0] = gc_images[i];
        } else if (color == "green") {
            canais[1] = gc_images[i];
        } else if (color == "red") {
            canais[2] = gc_images[i];
        } else {
            return gc_images; // Retorna cinza por padrão
        }

        cv::merge(canais, colored_images[i]);
    }
    return colored_images;
}

void GrayCode::set_px_f(int pixel_f){
    px_f = pixel_f;

    gc_images.clear();
    n_bits = static_cast<int>(std::ceil(std::log2(static_cast<double>(width) / px_f) + 1.0));
    gc_images.resize(n_bits+2);

    for (int i = 0; i < n_bits+2; ++i) {
        gc_images[i] = cv::Mat::zeros(height, width, CV_8UC1);
    }
}

File: ros2_active_stereo/include/gray_code.py
import numpy as np
import cv2


class GrayCode:
    def __init__(self, resolution=(512, 512), n_bits=4, axis=0, px_f=16):
        """
            Inicializa uma instância para a criação de imagens de Gray Code.

            Este metodo configura os parâmetros necessários para gerar imagens de Gray Code e inicializa
            a matriz de imagens com zeros. Ele também cria as imagens de Gray Code com base na resolução
            e no número de bits especificados.
        """
        self.width = resolution[0]
        self.height = resolution[1]
        self.px_f = px_f
        self.n_bits = n_bits  # Number of bits for the Gray code, adjusted to start from 0
        self.gc_images = np.zeros((self.height, self.width, self.n_bits + 2), np.uint8)  # Combined image sequence
        # self.create_images(axis=axis)  # Generate the Gray code images based on the specified axis
        self.create_graycode_images()

    def get_gc_images(self):
        """
            Retorna as imagens de Gray Code geradas.
        """
        return self.gc_images

    def show_image(self):
        """
            Exibe cada padrão de imagem na matriz `gc_images` em uma janela de visualização.
        """
        for i in range(self.gc_images.shape[2]):
            print(i)
            cv2.imshow('Image', self.gc_images[:, :, i])
            cv2.waitKey(0)
        cv2.destroyAllWindows()

    def create_graycode_images(self):
        """
            Gera imagens de código de Gray para codificação de padrões projetados em uma cena.

            Esta função cria uma sequência de imagens codificadas em Gray (Gray code) usadas em técnicas como
            projeção de padrões estruturados. Essas imagens são utilizadas para calcular a profundidade ou mapear
            a geometria da cena.

            Parameters:
            -----------
            Não possui parâmetros de entrada.

            Returns:
            --------
            Nenhum valor é retornado explicitamente. A função modifica o atributo `gc_images` da classe, que é
            uma matriz tridimensional de forma (altura, largura, canais) onde os padrões de código de Gray são
            armazenados.
        """
        n_bits = np.ceil(np.log2(self.width / self.px_f / 2))
        width = [element for element in np.arange(2 ** self.n_bits, dtype=np.uint8) for _ in range(int(self.px_f / 2))]
        width_b = self.list_to_graycode_binary(width, self.n_bits)
        self.gc_images[:, :, 0] = 255

        for j in range(self.width):
            for i in range(self.n_bits):
                self.gc_images[:, j, i + 2] = int(list(width_b[j])[i]) * 255

    def list_to_graycode_binary(self, int_list, bit_length):
        """
            Converte uma lista de inteiros em suas representações binárias em código de Gray.

            Esta função recebe uma lista de inteiros e converte cada inteiro em seu código de Gray
            correspondente. O código de Gray é uma forma de codificação binária onde dois números
            consecutivos diferem em apenas um bit.

            Parameters:
            -----------
            int_list : list of int
                Uma lista de inteiros que serão convertidos para código de Gray.

            bit_length : int
                O comprimento em bits que cada código de Gray deve ter. Este valor determina o
                número de dígitos binários na representação final, incluindo os zeros à esquerda.

            Returns:
            --------
            graycode_list : list of str
                Uma lista de strings, onde cada string é a representação binária do código de Gray
                correspondente a um inteiro da lista de entrada, com zeros à esquerda conforme
                especificado pelo `bit_length`.
        """
        graycode_list = []
        for n in int_list:
            graycode = n ^ (n >> 1)
            # Convert to binary string with leading zeros based on the bit length
            graycode_binary = format(graycode, f'0{bit_length}b')
            graycode_list.append(graycode_binary)
        return graycode_list

    def get_gc_order_v(self):
        """
            Obtém a ordem dos valores de código de Gray na imagem horizontal.

            Esta função analisa a porção relevante de uma imagem que contém padrões de código de Gray,
            converte esses padrões em valores binários, e então os converte em inteiros. Os valores únicos
            são identificados, indexados, e ordenados para determinar a sequência de código de Gray.

            Parameters:
            -----------
            Não possui parâmetros de entrada.

            Returns:
            --------
            sorted_qsi_val : numpy.ndarray
                Um array contendo os valores únicos de código de Gray, ordenados de acordo com sua
                posição original na imagem.
        """
        # Converter a porção relevante da imagem para valores binários
        bit_values = (self.gc_images[0, :, 2:] / 255).astype(int)
        # Converter os valores binários em inteiros
        int_values = np.dot(bit_values, 2 ** np.arange(bit_values.shape[-1])[::-1])
        # Obter valores únicos e seus índices
        qsi_val, indices = np.unique(int_values, return_index=True)

        # Combinar os valores únicos e índices em uma lista de tuplas e ordenar
        sorted_indices = np.argsort(indices)
        sorted_qsi_val = qsi_val[sorted_indices]

        return sorted_qsi_val